### PR TITLE
Update Ruff config

### PR DESCRIPTION
**Pull request by jarrodmillman**
## Description

1. Update pyproject.toml to use ruff's new linter settings (the old way has been deprecated).
2. Use project.requires-python  instead of tool.ruff.target-version as recommended by ruff.
3. Use NPY201 linter rule (hopefully this would have caught things like #7386.
4. Remove `line-length` and `exclude`.

## Checklist

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

We use [changelist](https://github.com/scientific-python/changelist) to
compile each pull request into an item of the release notes. Please refer to
the [instructions](https://scikit-image.org/docs/stable/development/contribute.html#documenting-changes)
and [past release notes](https://scikit-image.org/docs/stable/release_notes/index.html)
for guidance and examples.

```release-note
...
```

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -153,28 +153,30 @@ skip-string-normalization = 1
...
 line-length = 88
-target-version = 'py310'
+exclude = [
+    '.git',
+    '.ruff_cache',
+    'build',
+    'build-install',
+    'dist',
+    'doc/source/auto_examples',
+]
```
Comment (lagru) [inline pyproject.toml:163]: Why is this necessary? Since we only use ruff via pre-commit hooks I thought, that we can rely on `.gitignore` to cover this?

**Review by lagru — COMMENTED** [review]
Happy with all the changes here except for the one noted.

File: pyproject.toml (hunk 1 above)
Comment (jarrodmillman) [reply pyproject.toml:163]: It was added in #6729. I am not sure why. 

According to https://docs.astral.sh/ruff/settings/#exclude, `.git` and `.ruff_cache` are ignored by default and `build`, `build-install`, `dist`, and `doc/source/auto_examples/**/*` are in `.gitignore`.

I can try deleting and see what happens.

File: pyproject.toml [hunk 2]
```diff
@@ -153,28 +153,30 @@ skip-string-normalization = 1
 
 [tool.ruff]
 line-length = 88
```
Comment (jarrodmillman) [inline pyproject.toml:155]: If we are removing `exclude`, we might as well remove `line-length` since it is also redundant.

File: pyproject.toml (hunk 1 above)
Comment (lagru) [reply pyproject.toml]: Ah, my bad. I didn't notice that the exclude-block was just moved. I thought this PR added the block as new configuration and couldn't explain it. Sorry for the confusion. I'm happy to go with either way that works. Though, relying on `.gitignore` should be more robust and less repetition.

**Review by lagru — APPROVED** [review]
Approving on the condition that the CI is happy. :)

File: pyproject.toml (hunk 1 above)
Comment (jarrodmillman) [reply pyproject.toml]: And now `pyproject.toml` is 9 lines shorter!
